test/doctest/Main.hs

```diff
@@ -7,17 +10,36 @@
 -- Stability   : experimental
 -- Portability : non-portable (GHC extensions)
 --
-
 module Main where
 
-import Build_doctests                           ( flags, pkgs, module_sources )
-import Data.Foldable                            ( traverse_ )
-import Test.DocTest
+-- import Build_doctests                           ( flags, pkgs, module_sources )
+-- import Data.Foldable                            ( traverse_ )
+-- import Test.DocTest
+
+-- main :: IO ()
+-- main = do
+--   traverse_ putStrLn args
+--   doctest args
+--   where
+--     args = flags ++ pkgs ++ module_sources
+
+import Data.Array.Accelerate.Control.Monad.Trans
+import Prelude ( IO, print )
 
 main :: IO ()
-main = do
-  traverse_ putStrLn args
-  doctest args
-  where
-    args = flags ++ pkgs ++ module_sources
+main = print bar
+
+-- App a ~ Exp Int -> Exp (Maybe (Either Bool (a, Int)))
+type App = StateT Int (ExceptT Bool (MaybeT Exp))
+
+bar :: Exp Int -> Exp (Maybe (Either Bool (Int, Int)))
+bar = runMaybeT . runExceptT . runStateT foo
 
+foo :: App Int
+foo = do
+  modify (+1)
+  _ <- get
+  i <- lift $ throwE False_ `catchE` match \case
+                                    False_ -> except $ Right_ 5
+                                    True_  -> except $ Left_ False_
+  return i
```

Comment:
> Good question. I'm not sure; do we have an example where we can use it? Can you use it in a way that is unsurprising (does the thing you would expect)?

This is the little example I came up with earlier, but it's obviously not a practical use-case. It does (I think) do what you would expect: It does the exact same thing as the non-accelerate equivalent would do (which you get by removing underscores and the `match` function), but lifted to Accelerate terms.
> If we do want this though, this shouldn't replace the basic definition of Monad etc. because it will be confusing to have Monad from accelerate and a different Monad from accelerate-mtl.

It's probably possible to merge `D.A.A.Control.Monad.Monad` with `D.A.A.Control.Monad.Trans.Monad`, but I'm not sure it would be any less confusing (I think the normal `Monad` would give way better error messages than the other one, for a much more common use case).

> I'm also not sure if we want the mtl/transformers style for composing monads. The problem with that approach being that you need to declare n^2 instances for MonadTrans, MonadState, etc... And typically we ignore this problem because we can get GHC to derive the instances automatically, but here I am pretty sure you'll have to write those by hand, and then we are not going to make any friends. So should we just skip straight to an algebraic-effects style library, and what does that look like?

Currently I've implemented (a subset of) `transformers`, which does not require n^2 instances but instead you have to explicitly lift the operations (for the example linked here, you'd have to `lift . lift` to use the `MaybeT` monad). Indeed, layering `mtl` on top would be a bunch of grunt work. I haven't dived deep enough into AE to know how easy they would be, but I think it helps that we don't care about its performance in the way those libraries usually do: We won't execute the AE in a tight loop, only traverse it once to build up our own AST.

> BTW you've included a bunch of *.hi and *.o files in the commit(s).

Oops! I'll check on my gitignore :)